ledger/txtail.go

```diff
@@ -71,19 +79,44 @@ func (t *txTail) loadFromDisk(l ledgerForTracker) error {
 			return err
 		}
 
+		consensusParams := config.Consensus[blk.CurrentProtocol]
 		t.recent[old] = roundTxMembers{
-			txids:    make(map[transactions.Txid]basics.Round),
-			txleases: make(map[ledgercore.Txlease]basics.Round),
-			proto:    config.Consensus[blk.CurrentProtocol],
+			txleases: make(map[ledgercore.Txlease]basics.Round, len(payset)),
+			proto:    consensusParams,
 		}
+
 		for _, txad := range payset {
 			tx := txad.SignedTxn
-			t.recent[old].txids[tx.ID()] = tx.Txn.LastValid
-			t.recent[old].txleases[ledgercore.Txlease{Sender: tx.Txn.Sender, Lease: tx.Txn.Lease}] = tx.Txn.LastValid
-			t.putLV(tx.Txn.LastValid, tx.ID())
+			if consensusParams.SupportTransactionLeases && (tx.Txn.Lease != [32]byte{}) {
+				t.recent[old].txleases[ledgercore.Txlease{Sender: tx.Txn.Sender, Lease: tx.Txn.Lease}] = tx.Txn.LastValid
+			}
+			if tx.Txn.LastValid > t.lowWaterMark {
+				list := roundsLastValids[tx.Txn.LastValid]
+				// if the list reached capacity, resize.
+				if len(list) == cap(list) {
+					var newList []transactions.Txid
+					if cap(list) == 0 {
+						newList = make([]transactions.Txid, 0, initialLastValidArrayLen)
+					} else {
+						newList = make([]transactions.Txid, len(list), len(list)*2)
+					}
+					copy(newList[:], list[:])
+					list = newList
+				}
+				list = append(list, tx.ID())
+				roundsLastValids[tx.Txn.LastValid] = list
+			}
 		}
 	}
 
+	// add all the entries in roundsLastValids to their corresponding map entry in t.lastValid
+	for lastValid, list := range roundsLastValids {
+		lastValueMap := make(map[transactions.Txid]struct{}, len(list))
+		for _, id := range list {
+			lastValueMap[id] = struct{}{}
+		}
+		t.lastValid[lastValid] = lastValueMap
```

Comment:
there is no `t.putLV` in new code, is it functionally equivalent?

Reply:
yes; it's doing exactly what `t.putLV` does, but pre-allocate the maps to exact size needed. This is very important when we start a node that was under a great transactional load before : we don't want the node to start up consuming lot's of memory that isn't really needed. Instead, it allocates *exactly* the amount of needed map space. Note that as a side affect, the `roundsLastValids` here could be substantial, however, it will all be dumped when this method aborts.